### DISCUSSION TITLE
Controlling `Expect` response with RestEasy Reactive

### DISCUSSION THREAD

**Discussion by kdubb** [quarkusio/quarkus#24782] [Q&A]
We have a case where we _sometimes_ need to control the response when the `Expect: 100-continue` header is present. For example, before receiving the entire request body we _might_ want to forward the upload to a new destination.

Is there a means to do this with RestEasy Reactive?

[a comment by an automated account is not shown]

**Comment by geoand** [comment]
Nothing exists out of the box.

Can you describe the scenario in a little more detail?

Thanks

**Comment by kdubb** [comment]
Basically we have uploads that _may_ be rather large. For smaller requests the uploaded data is stored locally in our database cluster but for larger uploads we store the data indirectly in S3.

So the goal is to receive the initial request and decide if we will store the data locally or in S3. If storing locally, obviously we want to return `100-continue` and read the request immediately after. When storing in S3, the request would be forwarded to a pre-signed S3 url.

This will allow us to offload large uploads completely to S3 without altering the API or requiring the decision logic to be pushed to the client.

**Reply by geoand** [reply]
I see, thanks for explaining. @cescoffier I assume Vert.x handled `100-continue`, right? If so, is there anything special in the way its done or any examples I can see?

**Reply by cescoffier** [reply]
I don't believe there is anything special, but the use case is quite interesting. I would be interested by @pmlopes opinion on it.

**Reply by pmlopes** [reply]
Vert.x (web included) can handle this, take a look at this reproducer:

https://github.com/vert-x3/vertx-web/blob/issues/continue/vertx-web/src/test/java/io/vertx/ext/web/Router100ContinueTest.java#L50-L61

I've detailed a bit more on the code in a comment bellow

**Comment by kdubb** [comment]
Since nothing currently exists, I was trying to determine a good "interface" that would fit with JAX-RS/RestEasy Reactive.

It seems like it _should_ be doable without any special annotations or interface changes. Just declaring a `Uni<byte[]>` (or equivalent reactive stream of bytes) as the body parameter seems to be enough. If an `Expect` header is present on a method declared with a reactive stream of bytes as the body parameter it could pass a special stream implementation that first sends the `100-continue` and then pipes the incoming request bytes downstream. 

For special cases (e.g. redirection), the resource method can throw an appropriate `WebApplicationException` to generate those responses.


So something along the lines of...
```java
@Path("/upload")
public class UploadsResource {

  @POST
  public void upload(Uni<byte[]> bodyStream, @HeaderParam("Content-Length") Integer contentLength) {

    if (contentLength != null && contentLength > LOCAL_STORE_THRESHOLD) {
      URI redirectLocation = s3PreSigner.presignGetObject(...);
      throw RedirectionException(Response.Status.OK, redirectLocation);
    }    

    storeRequest(bodyStream);
  }

}
```

**Comment by FroMage** [comment]
That's a very interesting use-case. This sounds a lot like a custom pre-condition. JAX-RS already has an API for standard HTTP pre-conditions: https://quarkus.io/guides/resteasy-reactive#preconditions

I wonder who handles the `100-continue` in our stack, if it's done automatically by Vert.x or if RESTEasy Reactive is supposed to do it by itself.

Perhaps in normal cases we want to handle it automatically (not even sure), and then disable the automatic handling with an annotation such as `@CustomPrecondition` on the endpoint, and let you do what you showed. 

I realise that our documentation mentions how to stream blocking request bodies, but not async ones: https://quarkus.io/guides/resteasy-reactive#accessing-the-request-body this is a bug on our part. I'm not even sure what async streaming we support for the request body. Do you know, @geoand @stuartwdouglas ?

**Reply by geoand** [reply]
I don't think we support any TBH...

**Reply by FroMage** [reply]
So what do we even do with the request body? Do we unconditionally read it into a file or what? Before we hand it off to the endpoint. Or does it depend on how the endpoint wants it?

**Reply by geoand** [reply]
We currently read it with `org.jboss.resteasy.reactive.server.handlers.FormBodyHandler` or `org.jboss.resteasy.reactive.server.handlers.InputHandler`

**Reply by FroMage** [reply]
I was going to ask why we go via a stupid `InputStream` and always defer reading to worker threads, but I guess this is because of the reader/interceptor API.

I suppose we could add support for `Uni<Entity>` which would allow sticking to the IO thread until we've read it all out and pass it to the reader/interceptor APIs.

Or additionally we could have an extension to `MessageBodyReader` to allow reading from Vert.x directly instead of `InputStream` but that would be a new async IO API.

This is all for async deserialising, though. If we only want async access to bytes/buffers and no reader/interceptor, we can easily support that and bypass `InputHandler` if the entity body is requested as an `AsyncStream` or `Multi<Buffer>` or whatever type. This would be enough for this particular use-case, and a good first start until someone wants async entity readers, no?

**Reply by geoand** [reply]
Yeah, I think so

**Reply by pmlopes** [reply]
Hi, vert.x can handle continue calls, here's a quick test I just added to verify it:

https://github.com/vert-x3/vertx-web/blob/issues/continue/vertx-web/src/test/java/io/vertx/ext/web/Router100ContinueTest.java#L47-L86

So the "business" logic just needs to be encapsulated in a handler like:

```java
    router.route()
      .handler((PlatformHandler) ctx -> {
        should.assertEquals("100-continue", ctx.request().getHeader(HttpHeaders.EXPECT));
        // Send a 100 continue response
        ctx.response().writeContinue();
        ctx.next();
      })
```

And then you just add the remaining handlers as usual, including the `BodyParser` (just like in the link above).

**Reply by FroMage** [reply]
OK, so it's not automatic. Thanks.

**Reply by pmlopes** [reply]
No, because there is no generic logic to handle this. For example, should we end all requests that don't include the `EXPECT: 100-continue` with `417` as described in the HTTP RFC?

Like the OP question, `if content length > XYZ` redirect, else `100` continue?

So, we can't assume anything, yet you can define a well-known behavior in Quarkus if you like, yet I'm uncertain if there is a `one size fits all` solution for this.

**Reply by FroMage** [reply]
Sure, that wasn't criticism, that's pretty logical :)

**Reply by pmlopes** [reply]
Sorry if I sounded harsh, that wasn't my intention. I was just trying to expose the reason why.

**Reply by stuartwdouglas** [reply]
So in Undertow by default we would trigger this if the user attempted to read the request (e.g. InputStream.read()). If the user did not attempt to read then we would not send the continue.

**Reply by kdubb** [reply]
@stuartwdouglas That is basically the example I created above correct? If I never read from the stream and threw a redirection exception it would respond with that response, correct?

**Reply by pmlopes** [reply]
So @stuartwdouglas, @FroMage you're suggesting that in vertx-web, since we handle the body read from `BodyHandler` we could have this extra code:

```java
String expect = ctx.request().getHeader(EXPECT);
if (expect != null) {
  // EXPECT present
  if ("100-continue".equalsIgnoreCase(expect)) {
    // signal the client to continue
    ctx.response().writeContinue();
  } else {
    // If I understand the RFC correctly, for other kind of EXPECT we should return 417
    ctx.fail(417);
    return;
  }
}

// continue with the body handler as it is today...
```

I think this could be a nice improvement if it makes sense. It won't solve the OP OOTB, as it doesn't allow custom logic to be run, however, if the `BodyHandler` has a `maxBodyLimit` set, one can try to capture the error that is "thrown" when this validation fails which is currently sending a 413 status if i'm not mistaken.

**Reply by stuartwdouglas** [reply]
You could almost argue that it should go in 'io.vertx.core.http.HttpServerRequest#handler', so if you register a handler to read input (and the request is not paused) then it will automatically handle this for you, but doing that in body handler would also be an improvement.

**Reply by pmlopes** [reply]
I'm more inclined to the `BodyHandler` because, doing it at the `HttpServerRequest` handler assumes that the handler will actually handle the POST body, which isn't known at that level.

**Reply by FroMage** [reply]
OK, that would solve the issue for the 100-Continue, but there's still no way to read bytes async in a reactive way in RR ;)
